File: leetcode/python/901. 股票价格跨度.py
#!/bin/bash python
"""
901. 股票价格跨度

思路：
依旧是单调栈，区别在于单独调用，并且增加了全局的天数信息
"""

class StockSpanner:

    def __init__(self):
        """
        依旧是单调栈
        保存天数和值，(day, price)
        单调栈，寻找每个元素左侧最近的更大值
        返回两个元素的下标之差，如果没有找到则返回1
        单调递减栈，存储价格和下标（使用day代替）
        """
        self.day = 0
        self.st = []
        

    def next(self, price: int) -> int:
        self.day += 1
        while self.st and self.st[-1][1] <= price:
            self.st.pop()
        ret = self.day - (self.st[-1][0] if self.st else 0)
        self.st.append((self.day, price))
        return ret


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: leetcode/cpp/901. 股票价格跨度.cpp
// 这里自行定义了数据结构，感觉还不错

class StockSpanner {
public:
    typedef struct DataPair {
        int day;
        int price;
        DataPair(int day=0, int price=0) : day(day) , price(price) {}
    } DataPair;

    int day = 0;
    std::stack<DataPair> st;
    StockSpanner() {}
    
    int next(int price) {
        day += 1;
        while (!st.empty() && st.top().price <= price) {
            st.pop();
        }
        int ret = day - (st.empty() ? 0 : st.top().day);
        st.push(DataPair(day, price));
        return ret;
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */

// ===== 使用pair的版本 =====
class StockSpanner {
public:
    int day;
    stack<pair<int, int>> st;

    StockSpanner() {
        day = 0;
    }
    
    int next(int price) {
        while (!st.empty() && st.top().first <= price) {
            st.pop();
        }
        int ret = ++day - (st.empty() ? 0 : st.top().second);
        st.push(make_pair(price, day));
        return ret;
    }
};